xir/program.py

```diff
@@ -137,27 +137,55 @@ def use_floats(self) -> bool:
         return self._use_floats
 
 
+class ObservableFactor:
+    """Observable factor to be used in observable definitions.
+
+    Args:
+        name (str): the name of the factor
+        params ([Any]): the parameters of the factor
+        wires ([int]): the wires this factor acts on
+    """
+
+    def __init__(self, name: str, params: Params, wires: Sequence[Wire]) -> None:
+        self.name = name
+        self.params = list(params or [])
+        self.wires = tuple(wires or [])
+
+    def __str__(self) -> str:
+        wires = ", ".join([str(v) for v in self.wires])
+        if len(self.params) == 0:
+            return f"{self.name}[{wires}]"
+        else:
+            params = ", ".join([str(v) for v in self.params])
+            return f"{self.name}({params})[{wires}]"
+
+
 class ObservableStmt:
     """Observable statements to be used in observable definitions.
 
     Args:
         pref (Decimal, int, str): prefactor to the observable terms
-        terms (list): list of observables and the wire(s) they are applied to
+        factors (list[ObservableFactor]): list of observable factors
```

Comment:
Does this need to be a list? Or can we get away with making this a [`Sequence`](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) or something instead?